src/type/validate.js

```diff
@@ -584,29 +623,43 @@ function createInputObjectCircularRefsValidator(
 
     const fields = Object.values(inputObj.getFields());
     for (const field of fields) {
-      if (isNonNullType(field.type) && isInputObjectType(field.type.ofType)) {
-        const fieldType = field.type.ofType;
-        const cycleIndex = fieldPathIndexByTypeName[fieldType.name];
-
-        fieldPath.push(field);
-        if (cycleIndex === undefined) {
-          detectCycleRecursive(fieldType);
-        } else {
-          const cyclePath = fieldPath.slice(cycleIndex);
-          const pathStr = cyclePath.map((fieldObj) => fieldObj.name).join('.');
-          context.reportError(
-            `Cannot reference Input Object "${fieldType.name}" within itself through a series of non-null fields: "${pathStr}".`,
-            cyclePath.map((fieldObj) => fieldObj.astNode),
-          );
+      const fieldType = getNamedType(field.type);
+      if (isInputObjectType(fieldType)) {
+        const isNonNullField =
+          isNonNullType(field.type) && field.type.ofType === fieldType;
+        if (isNonNullField || !isEmptyValue(field.defaultValue)) {
```

Comment:
This really just adds a condition, implementing this [comment suggestion](https://github.com/graphql/graphql-spec/pull/793#discussion_r619765293)

I definitely need to write tests to ensure this is behaving as expected